### PR TITLE
Fixes #10731 - Allow matches to merge with default values

### PR DESCRIPTION
**Pull request by orrabin**

### CODE REVIEW

File: app/views/lookup_keys/_fields.html.erb [hunk 1]
```diff
@@ -51,8 +50,10 @@
       <%= checkbox_f(f, :merge_overrides, :onchange => 'mergeOverridesChanged(this)', :table_field => true,
                         :disabled => !f.object.supports_merge?, :size => "col-md-1", :label_size => "col-md-3",
                         :help_inline => popover("", _("Continue to look for matches after first find (only array/hash type)? Note: merging overrides ignores all matchers that use Puppet default."))) %>
+      <%= checkbox_f(f, :merge_default, :disabled =>  !f.object.merge_overrides, :size => "col-md-1", :table_field => true,
+                        :label_size => "col-md-3", :help_inline => popover('',_("Should the overrides be merged with the default value."))) %>
```
Comment (domcleal) [inline app/views/lookup_keys/_fields.html.erb]: The API description's better as it doesn't start with "Should", it simply says what it does ("Include default value..."), I'd say use it here too.

**Comment by orrabin** [comment]
@domcleal thanks, I fixed the help message.

File: app/views/api/v2/smart_class_parameters/main.json.rabl [hunk 2]
```diff
@@ -3,4 +3,4 @@ object @smart_class_parameter
 extends "api/v2/smart_class_parameters/base"
 
 attributes :description, :override, :parameter_type, :default_value, :use_puppet_default, :required, :validator_type, :validator_rule,
-           :merge_overrides, :avoid_duplicates, :override_value_order, :override_values_count, :created_at, :updated_at
\ No newline at end of file
+           :merge_overrides, :merge_default, :avoid_duplicates, :override_value_order, :override_values_count, :created_at, :updated_at
```
Comment (ares) [inline app/views/api/v2/smart_class_parameters/main.json.rabl]: nit: missing empty line

File: test/unit/classification_test.rb [hunk 3]
```diff
@@ -439,6 +439,32 @@ def setup
...
+                             :override => true, :key_type => 'hash', :merge_overrides => true, :merge_default => true,
+                             :default_value => { :default => 'default' }, :path => "organization\nlocation",
+                             :puppetclass => puppetclasses(:one))
+
+    as_admin do
+      LookupValue.create! :lookup_key_id => key.id,
+                          :match => "location=#{taxonomies(:location1)}",
+                          :value => {:example => {:a => 'test'}},
+                          :use_puppet_default => false
+    end
```
Comment (ares) [inline test/unit/classification_test.rb]: I think both creations can be in one `as_admin` block

Comment (ares) [inline test/unit/classification_test.rb]: could you please apply the same in previous test and add default value there so we test, that default value gets overriden unless merge_default is true?

File: app/assets/javascripts/lookup_keys.js [hunk 4]
```diff
@@ -187,9 +187,11 @@ function keyTypeChange(item) {
 
 function mergeOverridesChanged(item) {
   var fields = $(item).closest('.fields');
-  var keyType = fields.find("[id$='_key_type']").val();
+  var keyType = fields.find("[id$='_key_type']").find("option").parent().val();
```
Comment (ares) [inline app/assets/javascripts/lookup_keys.js]: is this change related? I don't see where it comes from

File: app/services/classification/base.rb [hunk 5]
```diff
@@ -53,11 +53,12 @@ def values_hash(options = {})
         end
         value = nil
         if key.merge_overrides
+          default = key.default_value if key.merge_default
```
Comment (ares) [inline app/services/classification/base.rb]: I'd prefer being explicit here, this works since ruby saw the variable but looks like `default` might be uninitialized local variable after first look since it's hidden by if without else

``` ruby
default = key.merge_default ? key.default_value : nil
```

File: app/services/classification/base.rb [hunk 6]
```diff
@@ -178,11 +179,17 @@ def update_generic_matcher(lookup_values, options)
       computed_lookup_value
     end
 
-    def update_array_matcher(should_avoid_duplicates, lookup_values, options)
+    def update_array_matcher(default, should_avoid_duplicates, lookup_values, options)
       elements = []
       values = []
       element_names = []
 
+      if default.present?
```
Comment (ares) [inline app/services/classification/base.rb]: I think `unless default.nil?` would make more sense, what if default value is empty string?

File: app/services/classification/base.rb [hunk 7]
```diff
@@ -199,11 +206,17 @@ def update_array_matcher(should_avoid_duplicates, lookup_values, options)
       {:value => values, :element => elements, :element_name => element_names}
     end
 
-    def update_hash_matcher(lookup_values, options)
+    def update_hash_matcher(default, lookup_values, options)
       elements = []
       values = {}
       element_names = []
 
+      if default.present?
```
Comment (ares) [inline app/services/classification/base.rb]: same as above

File: app/assets/javascripts/lookup_keys.js (hunk 4 above)
Comment (orrabin) [inline app/assets/javascripts/lookup_keys.js]: It's not actually, just found a bug while testing and I fixed it to test my changes. 
I'll remove it from here and open it in a separate PR.
The key_type changed to select2 so now `val()` doesn't actually return the selected value so the checkboxes would be disabled until you saved the parameter as an array/hash.

**Comment by orrabin** [comment]
@ares thanks for the comments, I fixed and rebased

**Comment by ares** [comment]
[test]

File: db/migrate/20150811170401_add_merge_default_to_lookup_key.rb [hunk 8]
```diff
@@ -0,0 +1,5 @@
+class AddMergeDefaultToLookupKey < ActiveRecord::Migration
+  def change
+    add_column :lookup_keys, :merge_default, :boolean
```
Comment (ares) [inline db/migrate/20150811170401_add_merge_default_to_lookup_key.rb]: `:null => false` and `:default => false` would make sense I think

**Comment by ares** [comment]
One nit I found during testing, merge default checkbox is active even when I select integer (or any non-mergeable type) while other non relevant checkboxes are disabled.

Second thing (probably fixed by fixing ^) when the validation error is present, it's not displayed and the lookup key is not saved while no error message present, maybe related to http://projects.theforeman.org/issues/9981.

**Comment by ares** [comment]
As far as I understand, we don't have smart variable API in V1 so you added this only to V2, correct?

**Comment by orrabin** [comment]
@ares we have lookup_keys in V1 but they don't have merge overrides so I only added this to V2.
I fixed the comments and the error not showing is related to the bug you mentioned.

**Comment by ares** [comment]
Ok, makes sense. Have you opened the PR already? I don't remember the fix exactly and I'd like to test with it... Ideally I'd merge at once.

**Comment by orrabin** [comment]
@ares I intentionally didn't add merge_override to v1 since my understanding is unless we broke something we don't change it.

File: app/services/classification/base.rb [hunk 9]
```diff
@@ -178,11 +179,17 @@ def update_generic_matcher(lookup_values, options)
...
 
-    def update_array_matcher(should_avoid_duplicates, lookup_values, options)
+    def update_array_matcher(default, should_avoid_duplicates, lookup_values, options)
       elements = []
       values = []
       element_names = []
 
+      unless default.nil?
+        values += default
+        elements << _('default_value')
```
Comment (domcleal) [inline app/services/classification/base.rb]: This is shown in the "where did this value come from?" help text in the UI, isn't it?  If so, don't use an underscore as it's a user-visible string.  Either "default value" or perhaps use the model attribute context-sensitive style, `s_("LookupKey|Default value")`

File: test/unit/classification_test.rb [hunk 10]
```diff
@@ -410,7 +410,7 @@ def setup
 
   test 'smart variable of hash with merge_overrides and priority should return lookup_value hash with array of elements' do
     key = FactoryGirl.create(:lookup_key, :key_type => 'hash', :merge_overrides => true,
-                             :default_value => {}, :path => "organization\nos\nlocation",
+                             :default_value => {:default => 'example'}, :path => "organization\nos\nlocation",
```
Comment (domcleal) [inline test/unit/classification_test.rb]: What's this been added for?  To show that it doesn't change smart variable behaviour, or was it meant to actually change?

Comment (orrabin) [inline test/unit/classification_test.rb]: To show that it doesn't change merge_override's behavior when it's set to false, I'll divide into 2 tests.

File: app/services/classification/base.rb (hunk 9 above)
Comment (orrabin) [inline app/services/classification/base.rb]: thanks, I'll change to `s_("LookupKey|Default value")`

**Comment by orrabin** [comment]
@domcleal I added a test and changed the value for the UI help text

**Comment by domcleal** [comment]
Merged as 8347a5e, thanks @orrabin.
